### DISCUSSION TITLE
Plugin to save drafts to browser storage

### DISCUSSION THREAD

**Discussion by jussiarpalahti** [TiddlyWiki/TiddlyWiki5#5389] [Q&A] [answered]
Hi.

I was trying to find a plugin that saves drafts locally into the browser. My use case is mobile devices where web pages can be reloaded without my consent, thus losing any unsaved tiddler. I couldn't find such a plugin I'll try to write one. Since changes need to be persisted as they happen I don't want to use sync nor save. From within browser I can access any lost drafts if need be.

Is there already builtin events to track draft changes in Tiddlywiki I could subscribe to? Attaching a handler to text area node should also work with mutation observation to watch parent nodes.

Thank you.

**Comment by pmario** [comment, marked as answer]
There is a core plugin: **Browser Storage: Local storage in the browser**  which you can find in the default plugin loader

**Reply by jussiarpalahti** [reply]
Thank you. I didn't realise to search *within* Tiddlywiki plugins for this... 😅 

I also tried to find the aforementioned Tiddlywiki events list. https://tiddlywiki.com/dev/#Event%20Mechanism doesn't say much about this. I'll try with debugger next. Not that I just now need to do anything, but I'd like to learn.

**Reply by Jermolene** [reply]
Hi @jussiarpalahti walking through the code of the browser storage plugin might be a good place to start:

https://github.com/Jermolene/TiddlyWiki5/blob/master/plugins/tiddlywiki/browser-storage/startup.js

**Reply by jussiarpalahti** [reply]
Hi, @Jermolene.

Thanks, I read that code actually. I found handlers for events (https://github.com/Jermolene/TiddlyWiki5/blob/ca95f1069fae5223022ad3756366aeab79c9d48f/plugins/tiddlywiki/browser-storage/startup.js#L42). What I hadn't yet found was triggers for these events. And if there was a list where events such as "change" and "tm-clear-browser-storage" are explained.

Thus I was thinking I'd just see what code does and find the events through that.

**Reply by Jermolene** [reply]
The "tm-clear-browser-storage" is usually generated by a button; there's an example at https://github.com/Jermolene/TiddlyWiki5/blob/master/plugins/tiddlywiki/browser-storage/settings.tid#L13

The wiki "change" event is generated by the wiki store object:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/wiki.js#L166

**Reply by jussiarpalahti** [reply]
Ok. I see there's a message system using JS DOM events. Thanks. I'll dig deeper.
